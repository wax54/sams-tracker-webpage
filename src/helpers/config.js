export const timeFrames = {
    day: {
        title: "Every Day",
        val: 1,
        roundPlaces: 2
    },
    week: {
        title: "Every Week",
        val: 7,
        roundPlaces: 1
    },
    month: {
        title: "Every Month",
        val: 30,
        roundPlaces: 0
    }
}