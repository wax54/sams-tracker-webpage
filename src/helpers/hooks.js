import { useState, useEffect } from 'react';

function getWindowDimensions() {
    const { innerWidth: width, innerHeight: height } = window;
    return {
        width,
        height
    };
}

export function useWindowDimensions() {
    const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

    useEffect(() => {
        function handleResize() {
            setWindowDimensions(getWindowDimensions());
        }

        window.addEventListener('resize', handleResize);
        return () => window.removeEventListener('resize', handleResize);
    }, [setWindowDimensions]);

    return windowDimensions;
}


export function useFormFields(init={}) {
    const [formData, setFormData] = useState(init);
    
    const handleChange = evt => {
        const {name, value} = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    };
    const resetForm = () => setFormData(init);
    return [formData, handleChange, resetForm];
}