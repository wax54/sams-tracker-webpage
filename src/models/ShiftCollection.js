/**
 * A shift class, representing time spent at a certain date on a thing for a reason
 * 
 */
class Shift {
    /**
     * 
     * @param { Date } start 
     * @param { string } type 
     * @param { string } category  
     * @param { Date } stop
     * @param { Number } id
     * @param { Number } u_id
     */
    constructor({ start, type, category, stop, id, u_id }) {
        this.id = id;
        this.u_id = u_id;
        if (start instanceof Date) this.start = start;
        else this.start = new Date(start);

        this.type = type;
        this.category = category;
        if (stop) {
            if (stop instanceof Date) this.stop = stop;
            else this.stop = new Date(stop);
        }
    }

    /**
     * returns the duration in Milliseconds
     */
    duration() {
        if (this.stop) return this.stop.getTime() - this.start.getTime();
        else return new Date().getTime() - this.start.getTime();
    }

    /**
    * returns the hours as a decimal that the shift took to complete
    */
    getHours() {
        return this.duration() / 3600000 //MS in an Hour
    }

    getFormattedDuration(specificity = 2) {
        const duration = this.duration();
        const durationFormatted = timeFormat(duration, specificity);
        return durationFormatted;
    }

    equals(test, strict=true) {
        if(typeof test !== 'object') return false;
        else if (this.id && test.id && strict) {
            if(this.id === test.id) return true;
            else return false; 
        } else {
            if (this.category === test.category) {
                if (this.type === test.type) {
                    if (this.start.getTime() === test.start.getTime()) return true;
                    if (!strict) return true;
                }
            }
            return false;
        }
    }

    /**
     * sets the stop of the shift
     *
     * @param { Date } stop the stop Date/Time that this shift should stop, if none
     *          specified it defaults to now.
     */
    clockOut(stop = new Date()) {
        this.stop = stop;
    }

    /**
     * stops the currents shift at splitDateTime and starts a new shift 
     *      at the same moment of the specified type and category
     * 
     * @param { Date } splitDateTime the time to clock out the original shift
     * @param { String } type the type that the newshift should take on 
     * @param { String } category the category the newshift take on, the same category if none specified
     * @returns { Shift } returns the newly created shift 
     */
    splitShift(splitDateTime, type, category = this.category) {
        const newShift = new Shift({start: splitDateTime, type, category, stop: this.stop})
        this.clockOut(splitDateTime);
        return newShift;
    }

}


class ShiftCollection {
    static SORT_PARAMS = {
        START: "start",
        STOP: "stop",
        CATEGORY: "category",
        TYPE: "type",
        ID: "id",
        UID: "u_id",
        ASCENDING: "ascending",
        DESCENDING: "descending"
    };

    constructor(...newShifts) {
        this.shifts = [];
        this.add(...newShifts);
    }

    shiftsBy(sortParam = ShiftCollection.SORT_PARAMS.START, direction = ShiftCollection.SORT_PARAMS.DESCENDING) {
        const shifts = [...this.shifts];
        if(direction === ShiftCollection.SORT_PARAMS.ASCENDING)
            return shifts.sort((shift1, shift2) => {
                if(sortParam === ShiftCollection.SORT_PARAMS.STOP){
                    let stop1 = shift1.stop || new Date();
                    let stop2 = shift2.stop || new Date();
                    return stop1 > stop2 ? 1 : -1;
                }

                return shift1[sortParam] > shift2[sortParam] ? 1 : -1;
            });
        else {
            return shifts.sort((shift1, shift2) => {
                if (sortParam === ShiftCollection.SORT_PARAMS.STOP) {
                    let stop1 = shift1.stop || new Date();
                    let stop2 = shift2.stop || new Date();
                    return stop2 > stop1 ? 1 : -1;
                }
                return shift2[sortParam] > shift1[sortParam] ? 1 : -1;
            });
        }
    }

    toString() {
        return 'A ShiftCollection of length ' + this.length() + ' ' + this.shifts;
    }
    //future add with shifts being an object of objects of arrays
    // tempState = { ...state };
    // category = action.payload.category;
    // type = action.payload.type;
    // if(tempState[category]) {
    //     if (tempState[category][type]) {
    //         let temp = tempState[category][type];
    //         tempState[category][type] = [...temp, action.payload];
    //     } else {
    //         tempState[category][type] = [action.payload];
    //     }
    // } else {
    // tempState[category] = { [type]: [action.payload] };
    add(...newShifts) {
        for (let shift of newShifts) {
            if (!(shift instanceof Shift)) {
                shift = new Shift(shift);
            }
            this.shifts.push(shift);
        }
    }
    getTotalHours() {
        return this.shifts.reduce((totalHours, shift) => {
            return totalHours + shift.getHours();
        }, 0);
    }

    remove(removedShift) {
        const shiftIndex = this.shifts.findIndex((shift) => {
            if (shift.category === removedShift.category)
                if (shift.type === removedShift.type)
                    if (shift.start === removedShift.start) return true;
            return false;
        });

        if (shiftIndex !== -1) {
            this.shifts.splice(shiftIndex, 1);
            return true;
        }
        else return false;
    }

    contains(someShift, strict = false) {
        if (this.find(someShift, strict)) return true;
        return false;
    }
    /**
     * 
     * @param {string} initialCategory the initial category name to select the shifts
     * @param {string} newCategory the name all shifts selected should be changed to
     * @returns {boolean} true on success, false if no shifts of category initialCategory are found
     */
    changeCategory(initialCategory, newCategory) {
        const shifts = this.category(initialCategory).shifts; //shift collection is returned, get the array out of it

        if (shifts.length) {
            shifts.forEach(shift => { shift.category = newCategory });
            return true;
        }
        return false;
    }

    changeType(initialtype, newType) {
        const shifts = this.type(initialtype).shifts; //shift collection is returned, get the array out of it
        //if it's not empty
        if (shifts.length) {
            shifts.forEach(shift => { shift.type = newType });
            return true;
        }
        //emtpy, ogType doesn't exist
        return false;
    }

    find(someShift, strict = false) {
        for (let shift of this.shifts) {
            if(shift.equals(someShift, strict)) return shift;
        }
        return false;
    }

    getCategories() {
        const categories = new Set();
        for (let shift of this.shifts) {
            categories.add(shift.category);
        }
        return [...categories];
    }

    getTypes() {
        const types = new Set();
        for (let shift of this.shifts) {
            types.add(shift.type);
        }
        return [...types];
    }

    getShiftsAfter(dateTime) {
        return new ShiftCollection(...this.shifts
            .filter(({ start }) => dateTime.getTime() < start.getTime()));
    }
    getShiftsBefore(dateTime) {
        return new ShiftCollection(...this.shifts
            .filter(({ start }) => dateTime.getTime() < start.getTime()));
    }

    getCurrShifts() {
        return new ShiftCollection(...(this.shifts.filter((shift) => !shift.stop)));
    }

    category(queryString) {
        return this.filter('category', queryString);
    }

    type(queryString) {
        return this.filter('type', queryString);
    }

    filter(field, query) {
        const matched = this.shifts.filter(shift => shift[field] === query);
        return new ShiftCollection(...matched);
    }

    length() { return this.shifts.length };
}

/**
 * returns a date representing x days away from dateTime 
 * 
 * @param {number} days the number of days from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} a new Date that is x hours from dateTime
 */
function daysFrom(days, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * 60 * 24 * days)); //1000ms * 60s * 60m * 24hrs * X days
}

/**
 * returns a day representing x hours away from dateTime 
 * 
 * @param {number} hours the number of hours from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} a new Date that is x hours from dateTime
 */
function hoursFrom(hours, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * 60 * hours)); //1000ms * 60s * 60m * X hours
}

/**
 * returns a day representing x mins away from dateTime 
 * 
 * @param {number} mins the number of mins from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} the Date that is x mins from dateTime
 */
function minsFrom(mins, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * mins)); //1000ms * 60s * X mins
}

/**
 *
 * @param { number } ms A duration represented in Milliseconds negative values are returned as positive durations
 * @param { number } specificity How many scales of time you would like returned the biggest is always returned
 *
 */
function timeFormat(ms, specificity) {

    const MSINSEC = 1000;
    const MSINMINUTE = 60000;
    const MSINHOUR = 3600000;
    const MSINDAY = 86400000;
    const MSINMONTH = 2628000000;
    const MSINYEAR = MSINMONTH * 12;



    let result = '';
    let i = 0;
    if (ms < 0) {
        result += '- ';
        ms = ms * -1;
    }

    const y = Math.floor(ms / MSINYEAR);
    if (y) {
        i++;
        ms = ms - (y * MSINYEAR);
        result += y + ' years ';
        if (i === specificity) {
            return result;
        }
    }


    const mth = Math.floor(ms / MSINMONTH);
    if (mth) {
        i++;
        ms = ms - (mth * MSINMONTH);
        result += mth + ' Months ';
        if (i === specificity) {
            return result;
        }
    }

    const d = Math.floor(ms / MSINDAY);
    if (d) {
        i++;
        ms = ms - (d * MSINDAY);
        result += d + ' Days ';
        if (i === specificity) {
            return result;
        }
    }

    const h = Math.floor(ms / MSINHOUR);
    if (h) {
        i++;
        ms = ms - (h * MSINHOUR);
        result += h + ' Hours ';
        if (i === specificity) {
            return result;
        }
    }

    const min = Math.floor(ms / MSINMINUTE);
    if (min) {
        i++;
        ms = ms - (min * MSINMINUTE);
        result += min + ' mins ';
        if (i === specificity) {
            return result;
        }
    }
    const s = Math.floor(ms / MSINSEC);
    if (s) {
        i++;
        ms = ms - (s * MSINSEC);
        result += s + ' seconds ';
        if (i === specificity) {
            return result;
        }
    }
    /*
    if (ms) {
        i++;
        result += ms + ' ms ';
        if (i == specificity) {
            return result;
        }
    }
    */
    return result;

}
function timeFormatFromHours(hrs, specificity) {
    const ms = hrs * 3600000;
    return timeFormat(ms, specificity);
}

//testing purposes #cleanUp
// const deli = new ShiftCollection(...JSON.parse(localStorage.records));

export { ShiftCollection, Shift, minsFrom, hoursFrom, daysFrom, timeFormatFromHours }