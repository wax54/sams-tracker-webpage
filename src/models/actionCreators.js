import UserApi from "./Api";
import {v4 as uuid} from "uuid";
import UploadQueue from "./UploadQueue";
import { store } from "./store";
import { SET_TIMEFRAME } from "./actionTypes";

/** Timeframes */
export function setTimeFrame(timeframeKey) {
    return { type: SET_TIMEFRAME, payload: timeframeKey };
}

/** SHIFTS */
export function startShift(type, category) {
    const tempId = uuid();
    const newShift = {id: tempId, start: new Date(), type, category};
    
    // add a command and shift/data to the upload queue
    UploadQueue.addToQueue({type:"addShift", data: newShift});
    //add the shift to the store
    return { type: "START_SHIFT", payload: newShift };
}

export function resetShifts() {
    return { type: "RESET_SHIFTS" };
}

export function loadShifts() {
    return async function (dispatch) {
        const resp = await UserApi.getShifts();
        if (resp.status === true) {
            dispatch({ type: "ADD_SHIFTS", payload: resp.shifts });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
            //TODO make a queue for failed shift uploads to be merged later
            //dispatch({ type: "ADD_SHIFT_TO_UPLOAD_QUEUE", payload: newShift })
        }
    }
}

export function refreshShifts() {
    return async function (dispatch) {
        dispatch(resetShifts());
        return await dispatch(loadShifts());
    }
}

export function updateShift(shift) {
    return { type: "UPDATE_SHIFT", payload: shift };
}

export function updateAShift(shift) {
    UploadQueue.addToQueue({ type: "updateShift", data: shift });
    return updateShift(shift);
}

export function endShift(id) {
    const stop = new Date();
    return clockOutAt(id, stop);
}

export function clockOutAt(shiftId, stop) {
    const shift = store.getState().shifts[shiftId];
    shift.stop = stop;
    UploadQueue.addToQueue({type: 'updateShift', data: shift})
    return updateShift(shift);
}

export function deleteShift(shiftId) {
    UploadQueue.addToQueue({type: 'deleteShift', data: shiftId});
    return { type: "DELETE_SHIFT", shiftId: shiftId };
}

/** ERRORS (depreciated) */
export function resetErrors() {
    return { type: "RESET_ERRORS" };
}


/** Users */
export function setUser(userData) {
    return { type: "SET_USER", payload: userData };
}
export function resetUser() {
    return async function (dispatch) { 
        dispatch({ type: "RESET_USER" });
        dispatch(resetShifts());
        dispatch(resetGoals());
    }
}
export function authorizeUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.login({username, password});
        if(resp.status === true)
            dispatch(setUser(resp.user));
            await dispatch(refreshShifts());
            await dispatch(refreshGoals());

        return resp;
    }
}

export function registerUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.register({ username, password });
        if (resp.status === true)
            dispatch(setUser(resp.user));
            dispatch(resetShifts());
            dispatch(resetGoals());
        return resp;
    }
}

/** GOALS */
export function addGoal({type, category, seconds_per_day}) {
    const newGoal = { seconds_per_day, type, category };
    UploadQueue.addToQueue({type: "addGoal", data: newGoal});
    return {type: "ADD_GOAL", payload: newGoal};
}

export function resetGoals() {
    return { type: "RESET_GOALS" };
}

export function loadGoals() {
    return async function (dispatch) {
        const resp = await UserApi.getGoals();
        console.log(resp);
        if (resp.status === true) {
            dispatch({ type: "LOAD_GOALS", payload: resp.goals });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
        }
    }
}

export function refreshGoals() {
    return async function (dispatch) {
        dispatch(resetGoals());
        return await dispatch(loadGoals());
    }
}

export function updateGoal({type, category, seconds_per_day}) {
    return { type: "UPDATE_GOAL", payload: {type, category, seconds_per_day} };
}

export function updateAGoal(goal) {
    UploadQueue.addToQueue({type: "updateGoal", data: goal});
    return updateGoal(goal);
}

export function deleteGoal({ type, category }) {
    UploadQueue.addToQueue({type: "deleteGoal", data: { type, category }})
    return { type: "DELETE_GOAL", payload: { category, type } };
}