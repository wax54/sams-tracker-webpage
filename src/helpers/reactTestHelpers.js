import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"



function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

export function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

export async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}
