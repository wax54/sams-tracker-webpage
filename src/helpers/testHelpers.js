import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"

function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}



export {
    renderWithRedux,
    renderAndWaitFor
}