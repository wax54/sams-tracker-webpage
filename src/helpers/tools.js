import { ShiftCollection, Shift, daysFrom } from "../models/ShiftCollection";
export function makeColor (shift) {
    const str = shift.type + shift.category;
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    let colour = '#';
    for (let i = 0; i < 3; i++) {
        const value = (hash >> (i * 8)) & 0xFF;
        colour += ('00' + value.toString(16)).substr(-2);
    }
    return colour;
}

export function round(num, decimals = 0) {
    if (decimals < 0) {
        decimals = decimals * -1;
        const divider = 10 ** decimals;
        num = Math.round(num / divider) * divider;
    }
    const multiplier = 10 ** decimals;
    return Math.round(num * multiplier) / multiplier;
}

// shifts 
        // {
        //     _hours: 17.3,
        //     _categories: ["school", "my well being"],
        //     _currShifts:[shift],
        //     school: { 
        //         _hours: 12, 
        //         _types:["coding"], 
        //         coding: { 
        //             _hours: 12, 
        //             _shifts: [shift, shift, shift] 
        //         }
        //     },
        //     "my well being": {
        //         _hours: 5.3,
        //         _types: ["resting", "painting"],
        //         resting: {
        //             _hours: 2.7,
        //             _shifts: [shift,  shift]
        //         },
        //         painting: {
        //             _hours: 2.6,
        //             _shifts: [shift]
        //         }
        //     },
        //  }

export function getShiftsByCategory(storeShifts, timeFrame) {
    const shifts = {
        _allShifts: function () {
            const shifts = new ShiftCollection();
            this._categories.entries().forEach(cat => {
                this[cat]._types.forEach(type => {
                    this[cat][type]._shifts.forEach(shift => {
                        shifts.add(shift);
                    });
                });
            });
            return shifts;
        },
        _currShifts: [],
        _hours: 0,
        _categories: new Set()
    };
    const earliestDateInTimeFrame = daysFrom(-timeFrame.val);

    for (let key in storeShifts) {
        const shift = new Shift(storeShifts[key]);
        //only shows shifts in timeFrame
        if (shift.stop < earliestDateInTimeFrame) continue;

        const hours = shift.getHours();
        if (!shift.stop) shifts._currShifts.push(shift);
        //if there is no entry for this category...
        if (!shifts._categories.has(shift.category)) {
            //make it
            shifts._categories.add(shift.category);
            shifts[shift.category] = { _hours: 0, _types: new Set() };
        }
        //shifts[shift.category] is garunteed at this point
        //if there is no entry for this type in the category...
        if (!shifts[shift.category]._types.has(shift.type)) {
            // make it
            shifts[shift.category]._types.add(shift.type);
            shifts[shift.category][shift.type] = { _hours: 0, _shifts: [] };
        }
        shifts[shift.category][shift.type]._shifts.push(shift);
        shifts[shift.category][shift.type]._hours += hours;
        shifts[shift.category]._hours += hours;
        shifts._hours += hours;
    }
    return shifts;
}