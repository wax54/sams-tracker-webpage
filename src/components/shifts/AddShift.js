import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startShift } from "../../models/actionCreators";
import InputAlert from "../InputAlert";

const AddShift = () => {
    let currShifts = useSelector(({ shifts }) => {
        const currShifts = [];
        for (let key in shifts) {
            if (!shifts[key].stop) currShifts.push(shifts[key]);
        }
        return currShifts;
    });
    
    const INIT = {type: "", category: ""};
    const [ formData, setFormData ] = useState(INIT);
    const [ inputAlerts, setInputAlerts ] = useState({});
    const dispatch = useDispatch();

    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData(data => ({...data, [name]: value}));
    }
    function addAlert(name, msg) {
        setInputAlerts(alerts => ({
            ...alerts,
            [name]: msg
        }));

        setTimeout(() => {
            setInputAlerts(alerts => ({
                ...alerts,
                [name]: undefined
            }));
        }, 3000);
    }

    function handleSubmit(evt) {
        evt.preventDefault();
        const type = formData.type.toLowerCase().trim();
        const category = formData.category.toLowerCase().trim();
        if(type && category) {
            if (currShifts.some(shift => shift.type===type && shift.category === category)) {
                addAlert('both', `Already clocked in to ${type} for ${category}`); 
                return;
            }
            setFormData(INIT);
            dispatch(startShift(type, category));
        } else {
            if(!type) addAlert('type', 'input must be filled');
            if(!category) addAlert('category', 'input must be filled');
        }
    }

    return (
        <form id="new-shift" className="col-xl border m-3 p-4 rounded shadow " onSubmit={handleSubmit}>
            <div className="row align-items-center justify-content-around">
                <h1 className="col-lg-3 col-xl-12 display-4 text-center">New Shift!</h1>

                <div className="col-12 col-lg-2 col-xl-5 m-2 mb-3">
                    <label htmlFor="type" className="form-label">Work-Type</label>
                    <div className="input-group">
                        <input type="text" 
                            className="form-control" 
                            id="type" 
                            name="type"
                            value={formData.type}
                            onChange={handleChange}
                            placeholder="Coding" 
                        />
                    </div>
                    {inputAlerts.type ? <InputAlert msg={inputAlerts.type} /> : null}
                </div>
                <div className=" col-lg-3 col-xl-5 m-2 mb-3">
                    <label htmlFor="category" className="form-label">Category</label>
                    <div className="input-group">
                        <div className="input-group-prepend">
                            <span className="input-group-text">For </span>
                        </div>
                        <input type="text" 
                            className="form-control" 
                            id="category" 
                            name="category"
                            value={formData.category}
                            onChange={handleChange}
                            placeholder="Fun" />
                    </div>
                    {inputAlerts.category ? <InputAlert msg={inputAlerts.category} /> : null}
                </div>

                <div id="clockin-buttons" className="col m-2 mb-3">
                    {inputAlerts.both ? <InputAlert msg={inputAlerts.both} /> : null}

                    <button type="submit" className="btn btn-secondary col-12 p-3 px-md-5 ">Clock In</button>
                </div>
            </div>
        </form>
    )
};
export default AddShift