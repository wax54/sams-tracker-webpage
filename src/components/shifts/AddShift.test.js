import { renderWithRedux, renderAndWaitFor } from '../../helpers/testHelpers';
import AddShift from './AddShift';
import userEvent from '@testing-library/user-event';
import { store } from '../../models/store';
import { resetAll, startShift } from '../../models/actionCreators';

beforeEach(() => store.dispatch(resetAll()));

test("doesn't blow up", () => {
    renderWithRedux(AddShift);
});

test("matchesSnapshot", () => {
    const { asFragment } = renderWithRedux(AddShift);
    expect(asFragment()).toMatchSnapshot();
});

test("inputs on page", () => {
    const { getByLabelText } = renderWithRedux(AddShift);

    const workTypeInput = getByLabelText("Work-Type");
    expect(workTypeInput).toBeInTheDocument();
    const categoryInput = getByLabelText("Category");
    expect(categoryInput).toBeInTheDocument();
});

test("Can Type into Inputs", async () => {

    const { getByLabelText, getByText, debug } = renderWithRedux(AddShift);
    const workTypeInput = getByLabelText("Work-Type");

    const categoryInput = getByLabelText("Category");

    userEvent.type(workTypeInput, "HELLO");
    userEvent.type(categoryInput, "GOODBYE");

    expect(workTypeInput.value).toEqual("HELLO");
    expect(categoryInput.value).toEqual("GOODBYE");

});

test("Submit lowercases inputs", async () => {
    const { getByLabelText, getByText } = renderWithRedux(AddShift,);

    const workTypeInput = getByLabelText("Work-Type");
    const categoryInput = getByLabelText("Category");

    userEvent.type(workTypeInput, "HELLO");
    userEvent.type(categoryInput, "GOODBYE");

    const clockInBtn = getByText("Clock In");

    userEvent.click(clockInBtn);

    const shifts = store.getState().shifts;
    const shiftId = Object.keys(shifts)[0];
    
    expect(shifts[shiftId].category).toEqual("goodbye");
    expect(shifts[shiftId].type).toEqual("hello");


});

test("Submit fails any empty category", async () => {
    const { getByLabelText, getByText, debug } = renderWithRedux(AddShift,);

    const workTypeInput = getByLabelText("Work-Type");
    const categoryInput = getByLabelText("Category");

    userEvent.type(workTypeInput,"empty category");
    userEvent.type(categoryInput, "");

    const clockInBtn = getByText("Clock In");

    userEvent.click(clockInBtn);

    expect(Object.keys(store.getState().shifts).length).toBe(0);
    expect(getByText('input must be filled')).toBeInTheDocument();


});
test("Submit fails any empty type", async () => {
    const { getByLabelText, getByText, debug } = renderWithRedux(AddShift,);

    const workTypeInput = getByLabelText("Work-Type");
    const categoryInput = getByLabelText("Category");

    userEvent.type(workTypeInput, "");
    userEvent.type(categoryInput, "empty type");

    const clockInBtn = getByText("Clock In");

    userEvent.click(clockInBtn);

    expect(Object.keys(store.getState().shifts).length).toBe(0);
    expect(getByText('input must be filled')).toBeInTheDocument();

});


test("Submit fails if already working type and category", async () => {
    store.dispatch(startShift("testtype", "testcategory"));

    const { getByLabelText, getByText, debug } = renderWithRedux(AddShift);

    const workTypeInput = getByLabelText("Work-Type");
    const categoryInput = getByLabelText("Category");


    userEvent.type(workTypeInput, "testType");
    userEvent.type(categoryInput, "testCategory");

    const clockInBtn = getByText("Clock In");

    userEvent.click(clockInBtn);

    expect(Object.keys(store.getState().shifts).length).toBe(1);
    expect(getByText(`Already clocked in to testtype for testcategory`)).toBeInTheDocument();

});