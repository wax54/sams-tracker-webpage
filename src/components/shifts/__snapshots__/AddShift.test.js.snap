// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchesSnapshot 1`] = `
<DocumentFragment>
  <form
    class="col-xl border m-3 p-4 rounded shadow "
    id="new-shift"
  >
    <div
      class="row align-items-center justify-content-around"
    >
      <h1
        class="col-lg-3 col-xl-12 display-4 text-center"
      >
        New Shift!
      </h1>
      <div
        class="col-12 col-lg-2 col-xl-5 m-2 mb-3"
      >
        <label
          class="form-label"
          for="type"
        >
          Work-Type
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            id="type"
            name="type"
            placeholder="Coding"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class=" col-lg-3 col-xl-5 m-2 mb-3"
      >
        <label
          class="form-label"
          for="category"
        >
          Category
        </label>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              For 
            </span>
          </div>
          <input
            class="form-control"
            id="category"
            name="category"
            placeholder="Fun"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="col m-2 mb-3"
        id="clockin-buttons"
      >
        <button
          class="btn btn-secondary col-12 p-3 px-md-5 "
          type="submit"
        >
          Clock In
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
